src/Development/IDE/Plugin/CodeAction.hs

```diff
@@ -1109,31 +1109,40 @@ rangesForBinding' _ _ = []
 --       import (qualified) A (..) ..
 --   Places the new binding first, preserving whitespace.
 --   Copes with multi-line import lists
-addBindingToImportList :: T.Text -> T.Text -> T.Text -> T.Text
-addBindingToImportList parent renderedBinding importLine = case T.breakOn "(" importLine of
-  (pre, T.uncons -> Just (_, rest)) ->
-    -- If the data type is in the import list wiouht the constructor, we should remove it and import it again
-    let rest' = case parent of
-          "" -> ", " <> rest
-          _ -> case T.breakOn parent rest of
-            (h, T.stripPrefix parent -> Just r) -> case T.uncons (T.dropWhile isSpace r) of
-              Just (')', _) -> ")" <> h <> r
-              Just ('(', xs) -> let imported = T.takeWhile (/= ')') xs in T.concat ["," ,imported , "), " , h , removeHeadingComma (T.tail (T.dropWhile (/= ')') r))]
-              _ -> "), " <> h <> r
-            _ -> "), " <> rest
-        binding' = (if T.null parent then id else T.init) renderedBinding
-     in removeTrailingComma $ T.concat [pre, "(", binding', rest']
-  _ ->
-    error $
-      "importLine does not have the expected structure: "
-        <> T.unpack importLine
+addBindingToImportList :: IdentInfo -> T.Text -> T.Text
+addBindingToImportList IdentInfo {parent = _parent, ..} importLine =
+  case T.breakOn "(" importLine of
+    (pre, T.uncons -> Just (_, rest)) ->
+      case _parent of
+        -- the binding is not a constructor, add it to the head of import list
+        Nothing -> T.concat [pre, "(", rendered, addCommaIfNeeds rest]
+        Just parent -> case T.breakOn parent rest of
+          -- the binding is a constructor, and current import list contains its parent
+          -- `rest'` could be 1. `,...)`
+          --               or 2. `(),...)`
+          --               or 3. `(ConsA),...)`
+          --               or 4. `)`
+          (leading, T.stripPrefix parent -> Just rest') -> case T.uncons (T.stripStart rest') of
+            -- case 1: no children and parentheses, e.g. `import A(Foo,...)` --> `import A(Foo(Cons), ...)`
+            Just (',', rest'') -> T.concat [pre, "(", leading, parent, "(", rendered, ")", addCommaIfNeeds rest'']
+            -- case 2: no children but parentheses, e.g. `import A(Foo(),...)` --> `import A(Foo(Cons), ...)`
+            Just ('(', T.uncons -> Just (')', rest'')) -> T.concat [pre, "(", leading, parent, "(", rendered, ")", rest'']
+            -- case 3: children with parentheses, e.g. `import A(Foo(ConsA),...)` --> `import A(Foo(Cons, ConsA), ...)`
+            Just ('(', T.breakOn ")" -> (children, rest''))
```

Comment:
`T.stripStart` doesn't strip comments, so

    import M(Foo({-Cons-}))` 

will fall through case 3 and unexpectedly become 

    import M(Foo(ConsB, {-Cons-})

Not your fault though, your code is as good as it can reasonably be. But the only fully correct way to make edits in the user AST is with ghc-exactprint

Reply:
Things become quite complicated involving comments:

* block comments are considered to be constructors
```haskell
import M(Foo({-Cons-}))
```
⇓
```haskell
import M(Foo(ConsB, {-Cons-})
```

* unexpected `-}` after we matched the parent
```haskell
import M({-Foo-}Foo)
```
⇓
```haskell
error "unknown case"
```

* line comments 
```haskell
  import M(
--Foo
  Foo)
```
⇓
```haskell
error "unknown case"
```

* ...

It's knotty to resolve these cases simply relying on `breakOn` logic. As you suggested, manipulating ASTs would be an appropriate way. I'll check `ghc-exactprint` out :)